pkg/utils/tsoutil/tso_dispatcher.go

```diff
@@ -0,0 +1,220 @@
+// Copyright 2023 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tsoutil
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/tikv/pd/pkg/errs"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+const (
+	maxMergeRequests = 10000
+	// DefaultTSOProxyTimeout defines the default timeout value of TSP Proxying
+	DefaultTSOProxyTimeout = 3 * time.Second
+)
+
+type tsoResp interface {
+	GetTimestamp() *pdpb.Timestamp
+}
+
+// TSODispatcher dispatches the TSO requests to the corresponding forwarding TSO channels.
+type TSODispatcher struct {
+	tsoProxyHandleDuration prometheus.Histogram
+	tsoProxyBatchSize      prometheus.Histogram
+
+	// dispatchChs is used to dispatch different TSO requests to the corresponding forwarding TSO channels.
+	dispatchChs sync.Map // Store as map[string]chan Request
+}
+
+// NewTSODispatcher creates and returns a TSODispatcher
+func NewTSODispatcher(tsoProxyHandleDuration, tsoProxyBatchSize prometheus.Histogram) *TSODispatcher {
+	tsoDispatcher := &TSODispatcher{
+		tsoProxyHandleDuration: tsoProxyHandleDuration,
+		tsoProxyBatchSize:      tsoProxyBatchSize,
+	}
+	return tsoDispatcher
+}
+
+// DispatchRequest is the entry point for dispatching/forwarding a tso request to the detination host
+func (s *TSODispatcher) DispatchRequest(
+	ctx context.Context, req Request, tsoProtoFactory ProtoFactory, doneCh <-chan struct{}, errCh chan<- error) {
+	val, loaded := s.dispatchChs.LoadOrStore(req.getForwardedHost(), make(chan Request, maxMergeRequests))
+	reqCh := val.(chan Request)
+	if !loaded {
+		tsDeadlineCh := make(chan deadline, 1)
+		go s.dispatch(ctx, tsoProtoFactory, req.getForwardedHost(), req.getClientConn(), reqCh, tsDeadlineCh, doneCh, errCh)
+		go watchTSDeadline(ctx, tsDeadlineCh)
+	}
+	reqCh <- req
+}
+
+func (s *TSODispatcher) dispatch(
+	ctx context.Context, tsoProtoFactory ProtoFactory, forwardedHost string, clientConn *grpc.ClientConn,
+	tsoRequestCh <-chan Request, tsDeadlineCh chan<- deadline, doneCh <-chan struct{}, errCh chan<- error) {
+	dispatcherCtx, ctxCancel := context.WithCancel(ctx)
+	defer ctxCancel()
+	defer s.dispatchChs.Delete(forwardedHost)
+
+	log.Info("create tso forward stream", zap.String("forwarded-host", forwardedHost))
```

Comment:
seem to delete `errHandling:`, we need to check carefully.

Reply:
We just moved errHandling. This is because we previously call getDelegateClient() here to get *grpc.ClientConn from the given forwarded host and did errHandling here. Now we call getDelegateClient() outside of s.tsoDispatcher.DispatchRequest(), does the error handling there, then pass the *grpc.ClientConn to s.tsoDispatcher.DispatchRequest(). We didn't skip the error handling and the code is clearer.